sklearn/datasets/openml.py

```diff
@@ -0,0 +1,517 @@
+import gzip
+import json
+import os
+import shutil
+from os.path import join, exists
+from warnings import warn
+
+try:
+    # Python 3+
+    from urllib.request import urlopen
+except ImportError:
+    # Python 2
+    from urllib2 import urlopen
+
+
+import numpy as np
+import scipy.sparse
+
+from sklearn.externals import arff
+from .base import get_data_home
+from ..externals.six import string_types, PY2
+from ..externals.six.moves.urllib.error import HTTPError
+from ..utils import Bunch, Memory
+
+__all__ = ['fetch_openml']
+
+_OPENML_PREFIX = "https://openml.org/"
+_SEARCH_NAME = "api/v1/json/data/list/data_name/{}/limit/1"
+_DATA_INFO = "api/v1/json/data/{}"
+_DATA_FEATURES = "api/v1/json/data/features/{}"
+_DATA_FILE = "data/v1/download/{}"
+
+
+def _open_openml_url(openml_path, data_home):
+    """
+    Returns a resource from OpenML.org. Caches it to data_home if required.
+
+    Parameters
+    ----------
+    openml_path : str
+        OpenML URL that will be accessed. This will be prefixes with
+        _OPENML_PREFIX
+
+    data_home : str
+        Directory to which the files will be cached. If None, no caching will
+        be applied.
+
+    Returns
+    -------
+    result : stream
+        A stream to the OpenML resource
+    """
+    if data_home is None:
+        return urlopen(_OPENML_PREFIX + openml_path)
+    local_path = os.path.join(data_home, 'openml.org', str(hash(openml_path)))
```

Comment:
A python hash is not discriminative enough; there is a chance of collision. Why have you gone down this path?

Reply:
I.e. was there something wrong with having a directory structure to match the API as long as there aren't query parameters?

Reply:
This was quick and dirty fix to make it work, I can turn it back though.

No, nothing was wrong in general. It just felt very unnatural to have something as arbitrary as whatever is the last part of the url being the file, and all other url parts being folders. e.g., api/v1/data/list/data_id/1/limit being all folders, and then a file called 1.. But I guess this shouldn't matter

Reply:
Actually, yes there is a problem. Consider the following 2 urls:
api/v1/data/list/data_id/1
api/v1/data/list/data_id/1/data_version/1

these would clash on the filesystem (because api/v1/data/list/data_id/1 is now used as a file and the second url wants to make a folder out of it) Of course there are fixes possible, e.g., suffixing all urls with an extension in the caching mechanism. Or keep using python hash. Which one do you prefer?